Sb data release

skybristol opened the pull request:
I took @dwief-usgs 's ideas in a somewhat different direction. The main things we need here are an implementation of ScienceBase search that incorporates our processing metadata to to create a snapshot of a result for a data release or some other cache of ScienceBase items and a little bit of help on building some of the interesting filters we need. When I looked at the full enumeration of values for both systemType and browseCategory, they include data release pointers but some other interesting items. I added a listing of potential values to the sciencebase.Search class and changed the function to allow for these values and set them up as numbered filters.

There are still some safeguards that we might want to add here to take care of potential issues. It's fully possible to blow things up with this function, request every item in ScienceBase, and try to build a massive list. But I think the basic pattern should work for now. Here's a simple test:

``
import bispy
sb_search = bispy.sciencebase.Search()
sb_search.search_snapshot(
    system_type="Data Release", 
    fields="title,body,files,webLinks", 
    q="Canis lupus"
)

skybristol commented:
I need to move on from this point and rework some of the other response data structures. I'm merging this work into master, and will revisit with @dwief-usgs when he is back in the office.

dwief-usgs commented:
@skybristol 

Took a look at code.  Seems to cover the basis and functions correctly. Glad you addressed the filters.. thanks.   A few thoughts.  

- As you mentioned we will have to think through max returned items but I think for current needs a max of 1000 shouldn't hinder anything.  

- We could use a similar approach for fields as you did with filters, supplying a list of acceptable fields.  Not sure we need to though because I think sciencebasepy module handles invalid fields by only passing valid fields...  

- I think there will be many cases where we only want to search "official" sciencebase data releases.  If that is the case, should we add a function specific to data release?  something like...  maybe at this point that isn't doing enough to make it worth it?
def search_data_release(self, q=None, fields="id"):
          result = search_snapshot(system_type='Data Release', q=q, fields=fields)

- If your worried about not catching potential issues why did you drop the try except?  I understand if we use that method we might want to use it across modules, but it at least helps reduce the likelihood of the module failing outright.

- Initial Testing:

Your supplied test fails as you call the sciencebase module but the module is named sb.  Switching to sb the functions correctly.  I tried a few variations and got expected results based on SB web app searching but did not test this with to many variations

skybristol commented:
I actually think we should feed this additional functionality back to the sciencebasepy project and not do anything with it here. What we've introduced is really functionality that should be built into that core package for ScienceBase so that we don't have to maintain it separately. I just haven't gotten around to looking at what that would entail. They need better ways of packaging multiple filters together and some other methods to enumerate what the possible values for these filters would be, tying them to explanations on how the values get assigned. I actually pulled the available values that I hard coded into the function via the MongoDB dump of every item that I obtained for another purpose.

I just don't see the value add to us abstracting further into very specific searches against ScienceBase at this point. We can show how it's done in the examples we are working up, but I don't think we need to provide specific functions. The same could potentially be said about some of the other stuff we are doing, though most of the modules in this package now offer a pretty dramatic abstraction from what you get going directly to the various REST or other interfaces in terms of structuring query parameters in logical ways and especially formatting the response structures for usability.

I dropped the try/except in this case, because it wasn't actually going to catch anything as structured. I'm not opposed to it. We just need to follow the rule of making sure that we look for, catch, and handle specific exceptions in specific ways and not just in general. As this thing is still in development, I generally like to have code run into real roadblocks and blow up so I can then deal with whatever the issue is. As we get the code hardened for more production use, we'll have to put specific exceptions in and handle them appropriately.

dwief-usgs commented:
all sounds good, I'll try running this for our full WLCI use case and report back any issues encountered
